### PR TITLE
Add data structures for multilinear extensions

### PR DESCRIPTION
**Pull request by tsunrise**
## Description

- Add multilinear extension (evaluation form of multilinear polynomial where the domain is boolean hypercube) data structures in `ark-poly` crate. 
  - create module `evaluations`, and rename `Evaluations` as `evaluations::univariate::UVEvaluations`
  - add trait `MultilinearExtension`, and two implemetors `DenseMultilinearExtension` and `SparseMultilinearExtension`
  - Add dependency `hashbrown`
- Add related benchmarks in `ark-poly-benches`
  - Add benchmarks for sparse and dense multilinear extensions

closes: #XXXX

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [x] Targeted PR against correct branch (main)
- [x] Linked to Github issue with discussion and accepted design OR have an explanation in the PR that describes this work.
- [x] Wrote unit tests
- [x] Updated relevant documentation in the code
- [x] Added a relevant changelog entry to the `Pending` section in `CHANGELOG.md`
- [x] Re-reviewed `Files changed` in the Github PR explorer

### CODE REVIEW

**Comment by tsunrise** [comment]
Not sure why, but there's some compilation problem related to serialization after I merged #144  today. e.g.: 
```
error[E0276]: impl has stricter requirements than trait
  --> poly\src\domain\radix2.rs:26:64
   |
26 | #[derive(Copy, Clone, Hash, Eq, PartialEq, CanonicalSerialize, CanonicalDeserialize)]
   |                                                                ^^^^^^^^^^^^^^^^^^^^ impl has extra requirement `R: ark_std::io::Read`
```
and
```
error[E0277]: the trait bound `W: ark_serialize::Write` is not satisfied
  --> poly\src\domain\general.rs:49:63
   |
49 |             GeneralEvaluationDomain::Radix2(d) => d.serialize(&mut writer),
   |                                                               ^^^^^^^^^^^ the trait `ark_serialize::Write` is not implemented for `W`
```
Is there any workaround? @weikengchen

**Comment by weikengchen** [comment]
Let me have a try

**Comment by weikengchen** [comment]
It seems that the cause is as follows:

- This PR uses `ark_std`'s `rand` branch.
- The `ark-serialize` uses the `master` branch of `ark_std`.

Is there a reason to use the `rand` branch? Otherwise, this PR may have a dependency on merging of the `rand` branch.

**Comment by tsunrise** [comment]
Thanks! I am using `ark-std::{UniformRand, test_rng}` because `ark_ff::test_rng` was marked deprecated. I will use `ark_std` master branch instead.

**Comment by tsunrise** [comment]
Hi Ryan, I have added trait `MVPolynomialCoefficientForm<F: Field>` and `MultilinearPolynomialEvaluationForm<F: Field>`, and `SparsePolynomial` implements `MVPolynomialCoefficientForm<F: Field>`. Does that make sense? @ryanleh

File: CHANGELOG.md
Comment (ValarDragon) [inline CHANGELOG.md:4]: Can you remove these two lines from here?

File: CHANGELOG.md
Comment (ValarDragon) [inline CHANGELOG.md:34]: Can you move this one entry up? (To make the numbers in sorted order)

File: poly-benches/benches/dense_multilinear.rs
Comment (ValarDragon) [inline poly-benches/benches/dense_multilinear.rs:1]: I think we should make the benchmark format style consistent between multilinear polynomials and `dense_polynomial` / `fft`. 

I prefer the approach of dense polynomial/fft since it reduces the boiler plate, and exposes flags for enabling particular benchmarks / using different fields.

**Review by ValarDragon — COMMENTED** [review]
Left comments on the changelog and arithmetic. Haven't reviewed the arithmetic 

Can you also update the `poly/README.md`?

File: poly-benches/benches/dense_multilinear.rs
Comment (ValarDragon) [inline poly-benches/benches/dense_multilinear.rs:55]: Any reason for this significance level / sample size as opposed to the default? Should we be using this everywhere?

File: poly/src/evaluations.rs
Comment (ValarDragon) [inline poly/src/evaluations.rs:1]:
```suggestion
//! A univariate polynomial represented in evaluations form.
```

Comment (ValarDragon) [reply poly/src/evaluations.rs:1]: Any thoughts on how would the API need to be changed to represent a multilinear polynomial in evaluation form? (Not needed for this PR)

File: poly/src/polynomial/mod.rs
Comment (ValarDragon) [inline poly/src/polynomial/mod.rs:95]:
```suggestion
    /// Relabel the point by swapping `k` scalars from positions `a..a+k` to positions `b..b+k`, and from position `b..b+k` to position `a..a+k` in vector.
```

File: poly/src/polynomial/multivariate/mod.rs
Comment (ValarDragon) [inline poly/src/polynomial/multivariate/mod.rs:173]: Can you add some more comments for what this function is doing

File: poly-benches/benches/dense_multilinear.rs
Comment (tsunrise) [reply poly-benches/benches/dense_multilinear.rs:55]: default config looks good. I will use default config instead

File: poly/src/evaluations.rs
Comment (tsunrise) [reply poly/src/evaluations.rs:1]: Not sure if it is the best solution, but my current idea is to let multilinear polynomial implement `MVPolynomial`, and `MVPolynomial` no longer has method `terms`, `from_coefficients...`. Those methods are moved to another trait called `MVPolynomialCoefficientForm` that extends `MVPolynomial`. Ryan's `SparsePolynomial` implements this trait. Similarly, there is a trait called `MultilinearPolynomialEvaluationForm` that also extends `MVPolynomial`. In this PR, I implemented `partial_evaluate`, which reduces the number of variables by evaluating part of the vector. There's also a handy method called `relabel`, which basically switches positions of scalars in a vector (by doing so, it is easy to change the order of argument of the polynomial). 

For future PRs, I'm thinking about whether we should let univariate evaluations implement `UVPolynomial` trait.

File: poly/src/polynomial/multivariate/mod.rs
Comment (tsunrise) [reply poly/src/polynomial/multivariate/mod.rs:173]: more comments added

File: poly/src/polynomial/mod.rs
Comment (tsunrise) [reply poly/src/polynomial/mod.rs]: resolved

File: poly-benches/benches/dense_multilinear.rs
Comment (tsunrise) [reply poly-benches/benches/dense_multilinear.rs]: Make sense. I removed this inline doc

File: CHANGELOG.md
Comment (tsunrise) [reply CHANGELOG.md:31]: resolved

File: CHANGELOG.md
Comment (tsunrise) [reply CHANGELOG.md]: resolved

**Comment by ValarDragon** [comment]
Re the readme, I meant could the poly crate readme be updated (https://github.com/arkworks-rs/algebra/blob/master/poly/README.md)

File: poly/src/polynomial/mod.rs
Comment (ValarDragon) [inline poly/src/polynomial/mod.rs:66]: Not really sure that it makes sense to have rand on the `MVPolynomial` trait, since the function doesn't really make sense in its current form for sparse polynomials, as you need more info to specify how sparse it is. (Ditto for `UVPolynomial`)

Maybe we should remove it from the trait here and in `UVPolynomial`? Thoughts @Pratyush @ryanleh 
(This can be done in a separate PR for removing from `UVPolynomial`)

File: poly/src/polynomial/mod.rs
Comment (ValarDragon) [inline poly/src/polynomial/mod.rs:93]: Why are these called partial points / partial evaluations? Is the idea that you take a multi-linear polynomial, fix one variable, and then get a `MultilinearPolynomialEvaluationForm` object?

**Review by ValarDragon — COMMENTED** [review]
Some more comments

File: poly/src/polynomial/mod.rs
Comment (tsunrise) [reply poly/src/polynomial/mod.rs:93]: Yes. Given a `PartialPoint` of dimension n, `partial_evaluate` will fix the first `n` variables and return a `MultilinearPolynomialEvaluationForm` object. Is it better to call this method `fix_variables` instead?

File: README.md
Comment (tsunrise) [reply README.md]: fixed

File: poly/src/polynomial/mod.rs
Comment (ValarDragon) [reply poly/src/polynomial/mod.rs:93]: Yeah, I think `fix_variables` is a better name. I'm confused as to why this is a separate struct though. E.g. why not make the API `poly.fix_variables({variable assignment vector}) -> MultilinearPolynomial`

Comment (tsunrise) [reply poly/src/polynomial/mod.rs:93]: I named it `MultilinearPolynomialEvaluationForm` because as the name suggests, the polynomial is represented by its evaluation over {0,1}^num_vars. Many optimizations (in sumcheck, multilinear polynomial commitment scheme, etc) on multilinear polynomial only works on the evaluation form. If I name it as `MultilinearPolynomial`, it may also include multilinear polynomial in coefficient form. If a multilinear polynomial struct is in coefficient form, then this polynomial can implement `MVPolynomialCoefficientForm` instead of `MultilinearPolynomialEvaluationForm`. 

Not sure though, but does `Multilinear Polynomial` usually means a multilinear extension of an array (i.e. the evaluation form)? If so, then I can name it as `MultilinearPolynomial` instead.

Comment (tsunrise) [reply poly/src/polynomial/mod.rs:93]: Also, I believe we can put `fix_variables` (Reduce the number of variables of the `self` by fixing the `partial_point.len()` variables at `partial_point`.) in `MVPolynomial` instead of `MultilinearPolynomialEvaluationForm` in future PR. Any idea? @ryanleh

Comment (ValarDragon) [reply poly/src/polynomial/mod.rs:93]: I see, I misunderstood what Evaluation form meant here. I don't understand why this is a trait not a struct though. How does a `MultilinearPolynomialEvaluationForm` trait differ from a `MultilinearPolynomialCoefficientForm` trait?

It'd be nice if we can have a consistent methodology for going between evaluations and coefficient representations for the univariate polynomials and multilinear polynomials.

Comment (ValarDragon) [reply poly/src/polynomial/mod.rs:93]: Like these concepts of partial evaluation, and relabeling seem like they apply for all multivariate polynomial in coefficient form

Comment (ValarDragon) [reply poly/src/polynomial/mod.rs:93]: Like these idea of partial evaluation, and relabeling seem like they apply for all multivariate polynomial in coefficient form

Comment (tsunrise) [reply poly/src/polynomial/mod.rs:93]: I made `MultilinearPolynomialEvaluationForm` a trait because there are two structs `DenseMultilinearPolynomial` and `SparseMultilinearPolynomial` that implements this trait. I shall probably rename it as `DenseMultilinearExtension` and `SparseMultilinearExtension`. 

If a multilinear polynomial struct is in coefficient form, I believe it should implement `MVPolynomialCoefficientForm` (if multilinear coefficient form has any special optimization, we can add it in future PR). 

For now, [univariate polynomial evaluation](https://github.com/arkworks-rs/algebra/blob/3856133923c148418ef21bc0eb456ba69185ca5f/poly/src/evaluations.rs#L18) form does not implement `Polynomial` trait. I think for this PR I should not let `MultilinearPolynomialEvaluationForm` extends `Polynomial` trait either. We can unify those APIs in later PR. What do you think?

Comment (ValarDragon) [reply poly/src/polynomial/mod.rs:93]: I think I'm now understanding what you mean. This PR is only implementing operations on the evaluations of a multi-linear polynomial. 

I'm still confused about this as the trait though. Don't the evaluations need an associated domain, to specify what points in the field the polynomial is being evaluated over? This is part of why If that was included, then I get having the evaluation form API. Also shouldn't the `relabel` and `fix variables` methods be pushed to `MVPolynomial`? (We can leave the implementation as stubs/panics in `MVSparsePolynomial` for now)

Also agreed re the univariate polynomial evaluation form. Maybe we should rethink it to be more along the lines of what you're doing in this PR.

Comment (tsunrise) [reply poly/src/polynomial/mod.rs]: I've done most of the refactoring, but I will check it again tomorrow. Once I finish, I will reopen this PR

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:1]:
```suggestion
//! Multilinear polynomial represented in dense evaluation form.
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:18]: What is `n` here?

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:20]: Maybe we should say `n = log_2(number_of_variables)`?

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:27]:
```suggestion
    /// Construct a new polynomial from a list of evaluations where the index 
    /// represents a point in {0,1}^`num_vars` in little endian form. For example, `0b1011` represents `P(1,1,0,1)`
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:34]:
```suggestion
    /// Construct a new polynomial from a list of evaluations where the index
    /// represents a point in {0,1}^`num_vars` in little endian form. For example, `0b1011` represents `P(1,1,0,1)`
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:40]:
```suggestion
            "The size of evaluations should be 2^num_vars."
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:36]:
```suggestion
        // assert that the number of variables matches the size of evaluations
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:57]:
```suggestion
            ark_std::mem::swap(&mut a, &mut b);
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:91]: Maybe we should change this to output `Option<F>` if the number of variables doesn't match?

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/mod.rs:45]: This comment is inaccurate now, no?

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/mod.rs:55]:
```suggestion
    /// Reduce the number of variables of `self` by fixing the `partial_point.len()` variables at `partial_point`.
```

File: poly/src/evaluations/multivariate/multilinear/sparse.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/sparse.rs:25]: Why is this needed?

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/dense.rs:18]: typo, should be {0,1}^`num_vars`. I'll fix that

File: poly/src/evaluations/multivariate/multilinear/sparse.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/sparse.rs:25]: `index` requires returning a reference, and the sparse list (i.e. the TreeMap) only contains nonzero elements, so `zero` serves as a dummy variable (see https://github.com/arkworks-rs/algebra/blob/366749faee1321c791b1a91ffa736e1bcff82aa6/poly/src/evaluations/multivariate/multilinear/sparse.rs#L209). Is there any better way?

File: CHANGELOG.md
Comment (ValarDragon) [inline CHANGELOG.md:30]: Per meeting, lets undo this until we find a unified API / trait definitions

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/dense.rs]: yes.

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/mod.rs]: yeah, I've changed the API and documentation now

File: poly/src/evaluations/multivariate/multilinear/sparse.rs
Comment (Pratyush) [reply poly/src/evaluations/multivariate/multilinear/sparse.rs:25]: I see. This is fine =)

File: README.md
Comment (Pratyush) [inline README.md:22]:
```suggestion
* [`ark-poly`](poly): Implements univariate, multivariate, and multilinear polynomials, and FFTs over finite fields. 
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:71]:
```suggestion
    /// Returns an iterator that iterates over the evaluations over {0,1}^`num_vars`
```

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/dense.rs:76]:
```suggestion
    /// Returns a mutable iterator that iterates over the evaluations over {0,1}^`num_vars`
```

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/mod.rs:17]:
```suggestion
/// This trait describes an interface for the multilinear extension 
/// of an array,
```

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (Pratyush) [inline poly/src/evaluations/multivariate/multilinear/mod.rs:18]:
```suggestion
/// The latter is a multilinear polynomial represented in terms of its evaluations over 
/// the domain {0,1}^`num_vars` (i.e. the Boolean hypercube).
```

**Review by Pratyush — APPROVED** [review]
LGTM modulo the documentation changes!

File: README.md
Comment (tsunrise) [reply README.md]: fixed

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/dense.rs]: fixed

File: poly/src/evaluations/multivariate/multilinear/dense.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/dense.rs]: fixed

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/mod.rs]: fixed

File: poly/src/evaluations/multivariate/multilinear/mod.rs
Comment (tsunrise) [reply poly/src/evaluations/multivariate/multilinear/mod.rs]: fixed

**Pull request merged**
